### DISCUSSION TITLE
"error_code": 400,

### DISCUSSION THREAD

**Discussion by ldpiuaeng** [strimzi/strimzi-kafka-bridge#696] [Q&A]
i hava done complete according to the tutorial, but a get 
{
    "error_code": 400,
    "message": "Bad Request"
}
https://github.com/strimzi/strimzi-kafka-bridge/blob/main/documentation/modules/proc-producing-messages-from-bridge-topics-partitions.adoc

**Comment by scholzj** [comment]
You should probably share the command you run, your Bridge configuration etc.

**Comment by ldpiuaeng** [comment]
this is my config：
#Bridge related settings
bridge.id=my-bridge
# uncomment one the following lines (bridge.tracing) to enable Jaeger tracing, check the documentation how to configure the tracer
# OpenTracing support
#bridge.tracing=jaeger
# OpenTelemetry support
#bridge.tracing=opentelemetry

#Apache Kafka common
kafka.bootstrap.servers=10.20.16.95:9092

#Apache Kafka producer
kafka.producer.acks=1

#Apache Kafka consumer
kafka.consumer.auto.offset.reset=earliest

#AMQP related settings
amqp.enabled=false
amqp.flowCredit=100
amqp.mode=SERVER
amqp.host=0.0.0.0
amqp.port=5672
amqp.certDir=
amqp.messageConverter=io.strimzi.kafka.bridge.amqp.converter.AmqpDefaultMessageConverter

#HTTP related settings
http.enabled=true
http.host=0.0.0.0
http.port=9080
#Enable CORS
http.cors.enabled=true
http.cors.allowedOrigins=*
http.cors.allowedMethods=GET,POST,PUT,DELETE,OPTIONS,PATCH


this is my rund commong :
curl -X POST   http://10.20.16.95:9080/topics/bridge-quickstart-topic   -H 'content-type: application/vnd.kafka.json.v2+json'   -d '{
    "records": [
        {
            "value": "sales-lead-0002",
            "partition": 2
        },
        {
            "value": "sales-lead-0003"
        }
    ]
}'

**Reply by scholzj** [reply]
Can you please format it properly as code to make it readable? Also, you have a valid response now to the command. So does it work now?

**Reply by ldpiuaeng** [reply]
sure 
this is my config：
#Bridge related settings
bridge.id=my-bridge
#uncomment one the following lines (bridge.tracing) to enable Jaeger tracing, check the documentation how to configure the tracer
#OpenTracing support
#bridge.tracing=jaeger
#OpenTelemetry support
#bridge.tracing=opentelemetry

#Apache Kafka common
kafka.bootstrap.servers=10.20.16.95:9092

#Apache Kafka producer
kafka.producer.acks=1

#Apache Kafka consumer
kafka.consumer.auto.offset.reset=earliest

#AMQP related settings
amqp.enabled=false
amqp.flowCredit=100
amqp.mode=SERVER
amqp.host=0.0.0.0
amqp.port=5672
amqp.certDir=
amqp.messageConverter=io.strimzi.kafka.bridge.amqp.converter.AmqpDefaultMessageConverter

#HTTP related settings
http.enabled=true
http.host=0.0.0.0
http.port=9080
#Enable CORS
http.cors.enabled=true
http.cors.allowedOrigins=*
http.cors.allowedMethods=GET,POST,PUT,DELETE,OPTIONS,PATCH

**Reply by ldpiuaeng** [reply]
this is my  request and the response :
![image](https://user-images.githubusercontent.com/41384109/201933598-3df87ac8-ca80-45f0-a61e-1332e5e0fb8c.png)

**Reply by scholzj** [reply]
What does the bridge log say?

**Reply by ldpiuaeng** [reply]
this ?
![image](https://user-images.githubusercontent.com/41384109/201934268-8c7bb730-1155-4c4d-b013-e668b6c6254e.png)

**Reply by scholzj** [reply]
Just share the full log I guess. Why is the client constantly disconnected? Does the connection to the Kafka broker work?

**Reply by ldpiuaeng** [reply]
i am sure that the kafka broker is aliving  .i  dont know why  the client constantly disconnected. 
![image](https://user-images.githubusercontent.com/41384109/201935216-4738cc83-1d45-416f-b6ac-4a7a89087f68.png)
and how to check the full log ?

**Reply by scholzj** [reply]
> and how to check the full log ?

That really depends on your enviroment and how you run it.

**Comment by autolyticus** [comment]
I've been running into the same issue:
```
$ curl http://kbridge.example.com/topics/los-reisub0-abc           (base) 211ms  Wed Dec  7 02:59:29 2022
{"name":"los-reisub0-abc","configs":{"compression.type":"producer","leader.replication.throttled.replicas":"","message.downconversion.enable":"true","min.insync.replicas":"1","segment.jitter.ms":"0","cleanup.policy":"delete","flush.ms":"9223372036854775807","follower.replication.throttled.replicas":"","segment.bytes":"1073741824","retention.ms":"604800000","flush.messages":"9223372036854775807","message.format.version":"3.0-IV1","max.compaction.lag.ms":"9223372036854775807","file.delete.delay.ms":"60000","max.message.bytes":"1048588","min.compaction.lag.ms":"0","message.timestamp.type":"CreateTime","preallocate":"false","min.cleanable.dirty.ratio":"0.5","index.interval.bytes":"4096","unclean.leader.election.enable":"false","retention.bytes":"-1","delete.retention.ms":"86400000","segment.ms":"604800000","message.timestamp.difference.max.ms":"9223372036854775807","segment.index.bytes":"10485760"},"partitions":[{"partition":0,"leader":0,"replicas":[{"broker":0,"leader":true,"in_sync":true}]}]}
```

So the connection is clearly fine, however, when attempting to publish some data on the producer
```
$ curl -XGET \
                                 http://kbridge.example.com/topics/los-reisub0-abc \
                                 -H 'Content-type: application/vnd.kafka.json.v2+json' \
                                 -d '{
                               "records": [
                                  {
                                      "key": "key-1",
                                      "value": "value-1"
                                  },
                                  {
                                      "key": "key-2",
                                      "value": "value-2"
                                  }
                              ]}'

{"error_code":400,"message":"Bad Request"}

```
Returns the same HTTP Status code 400. 

Here are the full logs doesn't look like there's anything too interesting after making the two above requests.
[kafka-bridge-bridge-7695dc98b6-t2j78.log](https://github.com/strimzi/strimzi-kafka-bridge/files/10172152/kafka-bridge-bridge-7695dc98b6-t2j78.log)


 I've deployed on on-prem K3S using Istio as the IngressGateway. Please let me know if you need any more details.

**Comment by ppatierno** [comment]
I think that this error could be related to this fix (on bridge 0.23.0) https://github.com/strimzi/strimzi-kafka-bridge/pull/714
My guess is that the content-type is not arriving as it should be to the bridge due to the gateway in the middle? Is that possible? I have never used Istio. So the thing is about checking what is really arriving to the bridge after the gateway.

**Reply by autolyticus** [reply]
Thanks for the reply, that's an interesting suggestion but no dice :( 

The header definitely seems to be forwarded, and I've verified with [Kubeshark](https://github.com/kubeshark/kubeshark) as well.

Request that's reaching the Ingress Gateway:
<img width="609" alt="image" src="https://user-images.githubusercontent.com/25099244/206164109-d2c0a8fd-a0ae-4ae3-8a98-d2a221c65c53.png">

Request that's passed to Kafka Bridge:

<img width="737" alt="image" src="https://user-images.githubusercontent.com/25099244/206164693-abe68f5c-6f19-41b5-88c6-3e1b191066da.png">

**Reply by ppatierno** [reply]
Are you able to try submitting the same HTTP request directly bypassing the Istio gateway? Just trying to remove each possibile cause one by one.

**Reply by autolyticus** [reply]
Yeah, I've tried that as well, by using `kubectl port-forward kafka-bridge-bridge-service`. Still the same error I'm afraid.

**Reply by autolyticus** [reply]
Is there any way to get more verbose logging from the Kafka Bridge?

I feel like this should definitely have more information somewhere.
```
2022-12-07 12:52:15 ERROR HttpBridge:541 - [2090170724] Request: from 127.0.0.1:42930, method = POST, path = /topics/los-reisub0-abc
2022-12-07 12:52:15 ERROR HttpBridge:578 - [2090170724] Response: statusCode = 400, message = Bad Request
```

**Reply by ppatierno** [reply]
Yes you should be able to do that by changing this log level in the log4j2.properties file:

logger.bridge.level = INFO

Anyway as I said before it could be related to a fix added in the 0.23.1 version. Can you try this latest version of the bridge please? That fix helps to get more info about this kind of errors when they happens in the Vert.x HTTP routing part and it's not strictly related to the bridge logic itself.

**Reply by autolyticus** [reply]
Just tried it today. I'm finding it even more confusing now :D

```bash
curl -X POST \
                                 http://<host>/topics/los-reisub0-abc \
                                 -H 'Content-type: application/vnd.kafka.json.v2+json' \
                                 -d '{
                               "records": [
                                  {
                                      "key": "key-1",
                                      "value": "value-1"
                                  },
                                  {
                                      "key": "key-2",
                                      "value": "value-2"
                                  }
                              ]
                          }' --http1.1
{"error_code":400,"message":"[Bad Request] Body application/vnd.kafka.json.v2+json parsing error: Null body"}
```

**Reply by autolyticus** [reply]
```
2022-12-10 16:12:40 ERROR HttpBridge:557 - [691417111] Request: from 127.0.0.6:42651, method = POST, path = /topics/los-reisub0-abc
2022-12-10 16:12:40 ERROR HttpBridge:595 - [691417111] Response: statusCode = 400, message = [Bad Request] Body application/vnd.kafka.json.v2+json parsing error: Null body
```
Just for the sake of completeness, that's the same message in the logs as well.

**Reply by ppatierno** [reply]
I agree the error is even more confusing and it's coming from the Vert.x HTTP routing part of the bridge :-(
The more interesting thing is that I just tried your same command on the bridge 0.23.1 locally and it works just fine for me.
Wondering what could be different in your environment.

**Reply by ppatierno** [reply]
I am not one of the Vert.x developers but it seems that your request is hitting this line https://github.com/vert-x3/vertx-web/blob/master/vertx-web-validation/src/main/java/io/vertx/ext/web/validation/impl/body/TextPlainBodyProcessorImpl.java#L30

Instead of using the `JsonBodyProcessorImpl` (https://github.com/vert-x3/vertx-web/blob/master/vertx-web-validation/src/main/java/io/vertx/ext/web/validation/impl/body/JsonBodyProcessorImpl.java), it's loading the `TextPlainBodyProcessorImpl` somehow as my understanding but I don't know why.
Even debugging the bridge locally it's using the JSON body processor and working fine.

@pmlopes ... as Vert.x web component expert do you have any idea why this kind of things could happen?

In all these tries, have you tried to run a single broker Kafka cluster locally on your laptop and the bridge as well? I would love to see Istio totally out of picture for the 0.23.1 version as well. Not sure if all those `X-***` headers are confusing the Vert.x HTTP routing somehow.

**Reply by autolyticus** [reply]
Yeah it's just bizarre. I keep hoping that I've made some silly mistake, but everything seems to be fine and instead of getting frustrated I'm getting very interested in finding out what the real reason is :D

As mentioned above: I've already tried port-forwarding the service from the cluster (which bypasses Istio completely). And just for the sake of completeness, I've even completely nuked Istio from my cluster, and replaced it with Contour. I was playing around with the Istio service mesh features before and I was wondering whether the [Istio sidecar](https://istio.io/latest/docs/setup/additional-setup/sidecar-injection/) might have something to do with it.

Contour is much simpler and doesn't have any of these fancy features and right now I just have a basic ingress set up. But still not working :D It's giving the same error.

```
❯ curl -X GET http://localhost:55599/topics                                                                                                                                                                                                ─╯
["__strimzi_store_topic","__strimzi-topic-operator-kstreams-topic-store-changelog","los-reisub0-abc"]
 ~/Work/k8s/strimzi  main ?59 ························································································································································································ base   21:27:08 ─╮
❯ curl -X POST http://localhost:55599/topics/los-reisub0-abc -H 'content-type: application/vnd.kafka.json.v2+json' -d '{"records": [{"key": "key-1","value": "value-1"},{"key": "key-2","value": "value-2"}]}'                                ─╯
{"error_code":400,"message":"[Bad Request] Body application/vnd.kafka.json.v2+json parsing error: Null body"}
```

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaBridge
metadata:
  name: kafka-bridge
spec:
  # ...
  image: quay.io/strimzi/kafka-bridge:0.23.1
  bootstrapServers: kafka-0-kafka-bootstrap:9092
  http:
    port: 8080
    cors:
      allowedOrigins: ["*"]
      allowedMethods: ["GET,POST,PUT,DELETE,OPTIONS,PATCH"]
  consumer:
    config:
      auto.offset.reset: earliest
  producer:
    config:
      delivery.timeout.ms: 300000
  # logging:
  #   type: inline
  #   loggers:
  #     logger.bridge.level: "INFO"
  template:
    pod:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
              - matchExpressions:
                - key: node-role.kubernetes.io/master
                  operator: In
                  values:
                  - "true"
```

Just for the sake of completeness, this is my KafkaBridge resource I'm using within `strimzi-cluster-operator`

The default configuration of a KafkaBridge resource seems to already have logging set to INFO so I don't think we can get more info than what we already have which is a shame :(

**Reply by ppatierno** [reply]
@reisub0 I think I found the problem. I didn't notice you enabled CORS when you posted the bridge configuration. Just noticed right now in the `KafkaBridge` custom resource. I was able to reproduce the problem. It seems that the producer side doesn't work but the other requests do. Investigating now ... I will keep you posted.

**Reply by ppatierno** [reply]
@reisub0 After some investigation and debugging some Vert.x internals I found an issue ... you can read more here https://github.com/strimzi/strimzi-kafka-bridge/issues/726
I will work on it to come to a fix.

**Reply by autolyticus** [reply]
Wow, thanks a lot for digging deep and finding this. Super interesting.

Never did I imagine CORS would have something to do with this (especially since I was not even using it through a browser!)

**Reply by ppatierno** [reply]
@reisub0 if you check on the issue, I opened a PR which works but still waiting for the Vert.x authors as well. Anyway, this fix will come in 0.24.0 or we could think about having a patch release 0.23.2 but not an ETA yet.
If you don't need CORS, the current release should work out of box.

**Reply by pmlopes** [reply]
Hi, I've commented on the fix PR. The fix is correct, though the vert.x API naming is misleading and should be addressed in the future.

**Reply by ppatierno** [reply]
@pmlopes I saw that! Thank you very much for jumping into this!

**Reply by autolyticus** [reply]
Sadly it doesn't seem to be possible to disable CORS through the `KafkaBridge` resource?

Both `spec.http.cors.allowedMethods` and `spec.http.cors.allowedRoutes` are [required values ](https://strimzi.io/docs/operators/latest/configuring.html#type-KafkaBridgeHttpConfig-reference)

**Reply by ppatierno** [reply]
Just remove the `cors` section under the `spec.http` one, it will do the trick.

**Reply by autolyticus** [reply]
Either way, the implementation aside, thanks a tonne for finding this, this was an interesting investigation. :D

**Reply by autolyticus** [reply]
> Just remove the `cors` section under the `spec.http` one, it will do the trick.

Nope, that's not possible, tried it already. The YAML won't pass validation when running `kubectl apply`
```
 kubectl explain kafkabridge.spec.http.cors                                                                                                                      (base) 2403ms  Mon Dec 12 19:49:31 2022
KIND:     KafkaBridge
VERSION:  kafka.strimzi.io/v1beta2

RESOURCE: cors <Object>

DESCRIPTION:
     CORS configuration for the HTTP Bridge.

FIELDS:
   allowedMethods       <[]string> -required-
     List of allowed HTTP methods.

   allowedOrigins       <[]string> -required-
     List of allowed origins. Java regular expressions can be used.
```

EDIT: Sorry! I just had to delete it first with `kubectl delete` and recreate the resource with `kubectl apply`. The validation check seems to be only when updating the resource.

**Reply by autolyticus** [reply]
Works fine without CORS :D

```
 curl -X POST \
http://<host>/topics/los-reisub0-abc \
-H 'content-type: application/vnd.kafka.json.v2+json' \
-d '{"records": [{"key": "key-1","value": "value-1"},{"key": "key-2","value": "value-2"}]}'
{"offsets":[{"partition":1,"offset":0},{"partition":1,"offset":1}]}⏎
```

**Reply by ppatierno** [reply]
Happy about that! You didn't need CORS in the end but thanks to your "mistake" we found a bug in the bridge :-D

**Reply by cgalmarini** [reply]
Hi, sorry for posting on this old issue but I'm having this same issue on 0.27 version. 
Any ideas about that?

**Reply by autolyticus** [reply]
I recommend opening a new issue with a detailed description of your setup and linking to this issue in case someone thinks it's relevant
